module/zfs/dmu_tx.c

```diff
@@ -334,6 +334,20 @@ dmu_tx_hold_write_by_dnode(dmu_tx_t *tx, dnode_t *dn, uint64_t off, int len)
 	}
 }
 
+void
+dmu_tx_hold_clone_by_dnode(dmu_tx_t *tx, dnode_t *dn, uint64_t off, int len)
+{
+	dmu_tx_hold_t *txh;
+
+	ASSERT0(tx->tx_txg);
+	ASSERT(len == 0 || UINT64_MAX - off >= len - 1);
+
+	txh = dmu_tx_hold_dnode_impl(tx, dn, THT_CLONE, off, len);
+	if (txh != NULL) {
+		dmu_tx_count_dnode(txh);
```

Comment:
I see that diverging from dmu_tx_hold_write_by_dnode() you've removed dmu_tx_count_write().  But aside of dropped write accounting, that is likely correct, you've also dropped layer-1 buffers pre-cache.  As I have written above I consider it ugly, but it supposed to reduce amount of waits and errors after transaction open.  You may check dmu_tx_hold_free_impl() for somewhat similar example.

Reply:
I had dmu_tx_count_clone() implementation initial based on dmu_tx_count_write(), but I was unable to trigger reading level-1 blocks (which I suspect is a bug).
Now I added dmu_tx_count_clone() that simply calls dmu_tx_hold_free_impl() (which I renamed to dmu_tx_count_free()). It does exactly what we need in case of offset and length being block-aligned.